std: sys: process: uefi: Use NULL stdin by default

Ayush1325 opened the pull request:
According to the docs in `Command::output`:

> By default, stdout and stderr are captured (and used to provide the
resulting output). Stdin is not inherited from the parent and any attempt by the child process to read from the stdin stream will result in the stream immediately closing.

This was being violated by UEFI which was inheriting stdin by default.

While the docs don't explicitly state that the default should be NULL, the behaviour seems like reading from NULL.

UEFI however, has a bit of a problem. The `EFI_SIMPLE_TEXT_INPUT_PROTOCOL` only provides support for reading 1 key press. This means that you either get an error, or it is assumed that the keypress was read successfully. So there is no way to have a successful read of length 0. Currently, I am returning UNSUPPORTED error when trying to read from NULL stdin. On linux however, you will get a read of length 0 for Null stdin.

One possible way to get around this is to translate one of the UEFI errors to a read 0 (Maybe unsupported?). It is also possible to have a non-standard error code, but well, not sure if we go that route.

Alternatively, if meaning of Stdio::Null is platform dependent, it should be fine to keep the current behaviour of returning an error.

cc @nicholasbishop @dvdhrm

Ayush1325 committed:
std: sys: process: uefi: Use NULL stdin by default

According to the docs in `Command::output`:

> By default, stdout and stderr are captured (and used to provide the
resulting output). Stdin is not inherited from the parent and any attempt
by the child process to read from the stdin stream will result in the
stream immediately closing.

This was being violated by UEFI which was inheriting stdin by default.

While the docs don't explicitly state that the default should be NULL,
the behaviour seems like reading from NULL.

UEFI however, has a bit of a problem. The `EFI_SIMPLE_TEXT_INPUT_PROTOCOL`
only provides support for reading 1 key press. This means that you
either get an error, or it is assumed that the keypress was read
successfully. So there is no way to have a successful read of length 0.
Currently, I am returning UNSUPPORTED error when trying to read from
NULL stdin. On linux however, you will get a read of length 0 for Null
stdin.

One possible way to get around this is to translate one of the UEFI
errors to a read 0 (Maybe unsupported?). It is also possible to have a
non-standard error code, but well, not sure if we go that route.

Alternatively, if meaning of Stdio::Null is platform dependent, it
should be fine to keep the current behaviour of returning an error.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

joboet commented:
> UEFI however, has a bit of a problem. The `EFI_SIMPLE_TEXT_INPUT_PROTOCOL` only provides support for reading 1 key press. This means that you either get an error, or it is assumed that the keypress was read successfully. So there is no way to have a successful read of length 0. Currently, I am returning UNSUPPORTED error when trying to read from NULL stdin. On linux however, you will get a read of length 0 for Null stdin.

You could use the `is_ebadf` logic to achieve this – [by making `is_ebadf` return `true` for `UNSUPPORTED`](https://github.com/rust-lang/rust/blob/c6c179662d5a6fc0520e05b5c0682dcfc7333f77/library/std/src/io/stdio.rs#L205), `read`ing from `Stdin` [will return `Ok(0)`](https://github.com/rust-lang/rust/blob/c6c179662d5a6fc0520e05b5c0682dcfc7333f77/library/std/src/io/stdio.rs#L100).

joboet commented:
> Currently, I am returning UNSUPPORTED error when trying to read from NULL stdin.

Wouldn't NOT_READY be closer to the intended meaning?

Ayush1325 commented:
> > Currently, I am returning UNSUPPORTED error when trying to read from NULL stdin.
> 
> Wouldn't NOT_READY be closer to the intended meaning?

Well, currently, stdio does retries on `NOT_READY`. I will be creating a new PR refactoring and fixing some things in Stdio so that we can use `NOT_READY` in that way. 

I also want to check if some keys register as 2 keys (currently Enter only registers as `\r` which means `read_line` does not work).

Ayush1325 commented:
@joboet I have created a PR to use `NOT_READY` in `is_ebadf`. I have also modified the Stdin code to avoid infinte loop.

Ayush1325 committed:
std: sys: process: uefi: Allow specifying Stdin

Stdio::MakePipe is not supported.

For Stdio::Null, return UNSUPPORTED. This is treated as read(0).
Additionally, have infinte loop on the notify function to prevent
wait_for_key from returning.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 commented:
@rustbot label +O-UEFI

joboet commented:
> > > Currently, I am returning UNSUPPORTED error when trying to read from NULL stdin.
> > 
> > 
> > Wouldn't NOT_READY be closer to the intended meaning?
> 
> Well, currently, stdio does retries on `NOT_READY`. I will be creating a new PR refactoring and fixing some things in Stdio so that we can use `NOT_READY` in that way.
> 
> I also want to check if some keys register as 2 keys (currently Enter only registers as `\r` which means `read_line` does not work).

Actually, I was wrong. `NOT_READY` implicates that waiting until the key event is signalled will result in a keypress being readable. For a null-handle, that never happens, so the key handle must never be signalled. But then waiting for a keypress will halt indefinitely. While of course `std` can work around this somehow, other programs probably won't.

I've reconsidered, and `UNSUPPORTED` is better. First of all, it's correct for the null-handle (it doesn't support being read from), but more importantly, it also makes sense that an `UNSUPPORTED` results in `Ok(0)` for actual devices as well – it matches what we do on platforms that don't have `Stdin`.

So, if you agree, please make both changes here – return `UNSUPPORTED` from `ReadKeyStroke` and detect it in `is_ebadf`.

Ayush1325 committed:
std: sys: stdio: uefi: Tread UNSUPPORTED Status as read(0)

Allows implementing Stdio::Null for Command in a deterministic manner.

Signed-off-by: Ayush Singh <ayush@beagleboard.org>

Ayush1325 commented:
> > > > Currently, I am returning UNSUPPORTED error when trying to read from NULL stdin.
> > > 
> > > 
> > > Wouldn't NOT_READY be closer to the intended meaning?
> > 
> > 
> > Well, currently, stdio does retries on `NOT_READY`. I will be creating a new PR refactoring and fixing some things in Stdio so that we can use `NOT_READY` in that way.
> > I also want to check if some keys register as 2 keys (currently Enter only registers as `\r` which means `read_line` does not work).
> 
> Actually, I was wrong. `NOT_READY` implicates that waiting until the key event is signalled will result in a keypress being readable. For a null-handle, that never happens, so the key handle must never be signalled. But then waiting for a keypress will halt indefinitely. While of course `std` can work around this somehow, other programs probably won't.
> 
> I've reconsidered, and `UNSUPPORTED` is better. First of all, it's correct for the null-handle (it doesn't support being read from), but more importantly, it also makes sense that an `UNSUPPORTED` results in `Ok(0)` for actual devices as well – it matches what we do on platforms that don't have `Stdin`.
> 
> So, if you agree, please make both changes here – return `UNSUPPORTED` from `ReadKeyStroke` and detect it in `is_ebadf`.

@joboet Hello, I have pushed the changes. 

Additionally, to make the wait for key press halt indefinitely, we need to add an infinite loop to the notify function. Is this what you are expecting?

joboet commented:
> Additionally, to make the wait for key press halt indefinitely, we need to add an infinite loop to the notify function. Is this what you are expecting?

No, I don't think that's right. If I understand UEFI events correctly, simply doing nothing in the notify-wait function and just never signalling the event should do the trick. That also makes it possible to wait on this event together with other (potentially signalled) events.

Ayush1325 commented:
> > Additionally, to make the wait for key press halt indefinitely, we need to add an infinite loop to the notify function. Is this what you are expecting?
> 
> No, I don't think that's right. If I understand UEFI events correctly, simply doing nothing in the notify-wait function and just never signalling the event should do the trick. That also makes it possible to wait on this event together with other (potentially signalled) events.

No, waiting on an event of type Wait will trigger the notify function for the event. Once the notify function concludes, the event is assumed to be signaled.
I have tested this behavior. If you add a print in the notify function, and remove the `is_ebadf`, and return `NOT_READY`, the notify function will keep on being triggered.

joboet commented:
> No, waiting on an event of type Wait will trigger the notify function for the event.

Yes, and this is documented behaviour ("this evaluation is repeated until an event is signaled or an error is detected", [7.1.5](https://uefi.org/specs/UEFI/2.10/07_Services_Boot_Services.html#efi-boot-services-waitforevent))

> Once the notify function concludes, the event is assumed to be signaled.

But this is not true: a `NOTIFY_WAIT` event's notification function being called is not the same as the event being signalled.

> I have tested this behavior. If you add a print in the notify function, and remove the `is_ebadf`, and return `NOT_READY`, the notify function will keep on being triggered infinitely.

Yes, and this is correct. But if you e.g. add a print after the `WaitForEvent`, that should never be reached.

Ayush1325 commented:
> Yes, and this is correct. But if you e.g. add a print after the WaitForEvent, that should never be reached.

I see. Seems for clearing up my misunderstanding. I have removed the infinite loop and tested that everything works fine.

@rustbot ready

joboet commented:
All right, let's ship it!
@bors r+

bors commented:
:pushpin: Commit d994fef7495ecbc1e18c95588253869f3a765edf has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet d994fef7495ecbc1e18c95588253869f3a765edf -->
<!-- homu: {"type":"Approved","sha":"d994fef7495ecbc1e18c95588253869f3a765edf","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
